ENH: calculation of region areas in spatial.SphericalVoronoi

pmla opened the pull request:
<!-- 
Thanks for contributing a pull request! Please ensure that
your PR satisfies the checklist before submitting:
http://scipy.github.io/devdocs/dev/contributor/development_workflow.html#checklist-before-submitting-a-pr

Also, please name and describe your PR as you would write a
commit message:
https://docs.scipy.org/doc/numpy/dev/gitwash/development_workflow.html#writing-the-commit-message

Note that we are a team of volunteers; we appreciate your
patience during the review process.

Again, thanks for contributing!
-->

#### What does this implement/fix?
<!--Please explain your changes.-->
This adds a method to calculate the areas of Voronoi regions in the `SphericalVoronoi` class.
Benchmarks:
```
============ =============
 num_points               
------------ -------------
    10          313±7μs   
    100       1.18±0.04ms 
    1000      9.66±0.2ms 
    5000      46.6±0.6ms 
   10000       93.6±2ms  
============ =============
```

#### Additional information
<!--Any additional information you think is important.-->
The `sort_vertices_of_regions` method sorts the regions in-place.  This means that we can't call it from within the `SphericalVoronoi` class without unexpected side-effects.  A call to the `calculate_areas` method added in this PR changes the region list.

pmla commented:
@tylerjereddy pinging you in here.

tylerjereddy commented on scipy/spatial/tests/test_spherical_voronoi.py:
```diff
@@ -170,3 +171,24 @@ def test_single_hemisphere_handling(self):
+                             itertools.product([0.5, 1, 2],
+                                               [(0, 0, 0), (1, 2, 3)]))
+    def test_equal_areas(self, radius, center):
+        points = radius * np.concatenate((-np.eye(3), np.eye(3))) + center
```
maybe add a comment here -- looks like you're checking the surface area footprint [projected from a cube](http://mathworld.wolfram.com/SolidAngle.html). Can you add a comment of this nature for clarity? And maybe a more specific test name to go with it?

pmla replied:
I have extended this test and renamed it `test_area_reconstitution_polyhedron`.  It now uses all the regular polyhedra as a sanity check.

tylerjereddy commented on scipy/spatial/_spherical_voronoi.py:
```diff
@@ -90,6 +106,10 @@ class SphericalVoronoi:
 
+    .. [VanOosterom] Van Oosterom and Strackee. The solid angle of a plane
+                     triangle. IEEE Transactions on Biomedical Engineering,
+                     2, 1983, pp 125--126.
```
The other thing is that manuscript isn't necessarily focused on calculating the surface areas of arbitrary convex spherical polygons, right? Reading through, it is focused on the spherical triangle / excess efficiency, which is nice, but I have some lingering concerns about generality, esp. given the far more recent concerns expressed in the [JPL paper](https://trs-new.jpl.nasa.gov/bitstream/handle/2014/40409/JPL%20Pub%2007-3%20%20w%20Errata.pdf?sequence=3&isAllowed=y)

tylerjereddy commented on scipy/spatial/_spherical_voronoi.py:
```diff
@@ -32,6 +32,17 @@ def sphere_check(points, radius, center):
+    denominator = 1 + np.einsum('ij,ij->i', R[:, 0], R[:, 1])\
+                    + np.einsum('ij,ij->i', R[:, 1], R[:, 2])\
+                    + np.einsum('ij,ij->i', R[:, 2], R[:, 0])
+    return np.abs(2 * np.arctan2(numerator, denominator))
```
I think that for all approaches to calculating spherical polygon surface areas there is some kind of weakness related to the accuracy of the spherical excess / angle calculations--here you don't have an interface to modulate the accuracy with i.e., discretization, which is a bit different than I did in #8069, so I'd like us to look into accuracy issues here---I'm happy for your approach to simply be better if it is :)

The authors of the manuscript you cite have clearly thought about this, and do note that there are 32 multiplications, 20 additions, 3 square roots, and 1 ATAN2. They also cite a discussion with a colleague at the University of Amsterdam about numerical stability.

tylerjereddy commented:
@rtshort if you have a minute to look things over here it may also be helpful to us

tylerjereddy commented:
One other thing -- let's not forget that if we do expand to N-dimensional spheres for SphericalVoronoi as  in https://github.com/scipy/scipy/pull/10692, then we may need guards of some sort on the area calculation to confine to a 3D sphere.

rtshort commented:
Will do.

pmla commented on benchmarks/benchmarks/spatial.py:
> Does pre-sorting actually make the area calculation any faster?

No - this is just me being silly.  I have removed it.

pmla commented on scipy/spatial/_spherical_voronoi.py:
```diff
@@ -32,6 +32,17 @@ def sphere_check(points, radius, center):
+    unit norm."""
+    numerator = np.einsum('ij,ij->i', R[:, 0], np.cross(R[:, 1], R[:, 2]))
+    denominator = 1 + np.einsum('ij,ij->i', R[:, 0], R[:, 1])\
+                    + np.einsum('ij,ij->i', R[:, 1], R[:, 2])\
```
Thanks. Fixed.

pmla commented on scipy/spatial/_spherical_voronoi.py:
```diff
@@ -65,6 +76,11 @@ class SphericalVoronoi:
+    Methods
+    ----------
+    calculate_areas
+        Calculates the areas of the Voronoi regions
```
>We may want to provide a brief description of the data structure (i.e., that the ordering of the areas in the returned array matches the ordering of the `regions` attribute)

I have described the shape of the output array, which is `(npoints,)`.  I think that it should be obvious that there is one region per input point, and that the areas calculated are in the same order.  Do you agree?

>Probably good to emphasize that there are spherical polygon areas on the surface of the sphere & not planar polygon areas.

I have added a note that the Voronoi regions are spherical polygons.

> A `versionadded` directive may also eventually be in order here

I have added a tentative `1.4.0` version.  We can update this later if it doesn't make the release.  I will leave this discussion unresolved so we don't forget.

pmla replied:
I have bumped `versionadded` to 1.5.0

pmla commented on scipy/spatial/_spherical_voronoi.py:
```diff
@@ -240,3 +257,50 @@ def sort_vertices_of_regions(self):
         _voronoi.sort_vertices_of_regions(self._tri.simplices, self.regions)
+
+    def calculate_areas(self):
+        """Calculates the areas of the Voronoi regions.
```
Done.

pmla commented:
> One other thing -- let's not forget that if we do expand to N-dimensional spheres for SphericalVoronoi as in #10692, then we may need guards of some sort on the area calculation to confine to a 3D sphere.

`sort_vertices_of_regions` in #10692 throws an exception if the input is not 3D, so in that respect we are safe.  I agree that it would be good to have the exception thrown in the `calculate_areas` function though.

pmla commented on scipy/spatial/_spherical_voronoi.py:
```diff
@@ -240,3 +257,50 @@ def sort_vertices_of_regions(self):
+        solid_angles = np.cumsum(triangle_solid_angles)[csizes - 1]
+        solid_angles[1:] -= solid_angles[:-1]
+
+        # get polygon areas using A = omega * r**2
```
Some thoughts on numerical accuracy:


> omega is basically just the spherical excess

Yes, they are the same.  From http://mathworld.wolfram.com/SphericalExcess.html

> The value of the excess is the solid angle (in steradians) subtended by the spherical triangle

In fact, for a sphere of radius 1, the spherical excess _is_ the spherical triangle area.  So, we should not discount the spherical excess per se.

It seems to me that the reason that the JPL approach is "clearly unusable" is that they have used an unnecessarily complicated expression for calculating the spherical excess.  They hint at this in the paper:

> ... it seems like there should be a simple expression for the area. If there is, we were unable to find it. A complicated exact answer can be found by the following sequence of steps ...

In their approach they have used l’Huiller’s formula, which at first glance appears to have very poor numerical stability.  I will boldly claim (without proper analysis) that the square root of the product of four tangent terms is not going to behave well.

In Van Oosterom and Strackee's method the highest order term inside the arctangent is a triple product, which should behave a lot better.

pmla commented on scipy/spatial/tests/test_spherical_voronoi.py:
```diff
@@ -170,3 +171,24 @@ def test_single_hemisphere_handling(self):
+                                               [(0, 0, 0), (1, 2, 3)]))
+    def test_areas(self, radius, center):
+        for points in [self.points, self.hemisphere_points,
+                       self.hemisphere_points2]:
```
Ah, nice. I didn't know you could do multiple parametrize calls.

pmla commented on scipy/spatial/tests/test_spherical_voronoi.py:
```diff
@@ -170,3 +171,24 @@ def test_single_hemisphere_handling(self):
+    @pytest.mark.parametrize("radius,center",
+                             itertools.product([0.5, 1, 2],
+                                               [(0, 0, 0), (1, 2, 3)]))
+    def test_areas(self, radius, center):
```
Thanks. Fixed.

pmla commented:
@tylerjereddy thanks for another great review. I'll reply to more of your comments later.

pmla commented:
> I wrote a few hundred lines of test code for spherical polygon areas in my previous detailed attempt to do this -- could you port some of those over to here to increase our confidence a bit?

I have added a test `test_area_reconstitution_large_input` which is similar to `test_functional_sph_Vor` from #8069.

I have looked over the other tests:

- `test_duplicate_filter`: duplicates already tested in `SphericalVoronoi` initialization.
- `test_half_hemisphere_area`, `test_half_hemisphere_area_reverse_order`: these are polygons rather than triangles.  The spherical triangles contained within the polygons should be covered by the "cube" polyhedron test though.
- `test_quarter_hemisphere_area`, `test_quarter_hemisphere_area_south`: these should be covered by the "cube" polyhedron test.
- `test_zero_radius_area`: a Voronoi region cannot have this shape.
- `TestSimplePlanarTri`: these tests contain points at the origin, which a Voronoi region cannot.
- `TestRadianAreas`: these should be covered by the "cube" polyhedron test.

I can't find any other tests which are applicable for Voronoi regions.  Do you have any particular tests in mind?

> This also reminds me that caculating surface areas of spherical polygons outside the context of the Voronoi diagram is also useful, so we may want to think about generality / API, or at least leaving the door open to that in the future?

Voronoi regions are very simple to triangulate.  Now that the single-hemisphere problem has been fixed, the triangulated regions have an unambiguous direction (they always face outwards from the origin).  Calculating surface areas or arbitrary spherical polygons does not have these nice properties.  So, other than the calculation of the triangle surface area, I'm not sure how much code would be shared between these applications.  In any case, the function `calculate_solid_angles` is not public, so I don't think that we have to worry about the API yet.

tylerjereddy commented:
> I can't find any other tests which are applicable for Voronoi regions. Do you have any particular tests in mind?

Definitely have a few in mind--would be nice to test some pathological cases where spherical polygons have edges that are almost parallel (small angles) or whatever else might probe numerical stability issues. The hard part can be knowing the expected value of the area in those cases. I think that's why I spent time writing a triangle subdivision algorithm in the other PR--to produce really small spherical triangles with known SAs.

Any degeneracy conditions? What happens if the input generators are just two antipodes? What about equi-spaced generators on the equator? I think we may not handle the last case robustly yet anyway, but if it does produce spherical rectangles that wrap around the sphere it might be pathological.

tylerjereddy commented on scipy/spatial/tests/test_spherical_voronoi.py:
```diff
@@ -170,3 +229,33 @@ def test_single_hemisphere_handling(self):
+    @pytest.mark.parametrize("radius", [0.5, 1, 2])
+    def test_area_reconstitution_large_input(self, radius):
+        np.random.seed(0)
+        n = 1000
```
bumping the number of generators to the point where the test fails (v. small Vor regions/ problem angles) might be something I should try locally & compare with discretization approach to see if the limits are similar

pmla replied:
I like this plan.

pmla replied:
I have tried random inputs containing up to 1 million points without failure.  I ran out of memory when trying 10 million input points.  In order to vectorize the code, I have to keep the triangulation of the Voronoi regions in memory.  If we want to reduce memory without sacrificing performance I think we will need to use Cython.

pmla commented:
> What happens if the input generators are just two antipodes?

Qhull will complain about this input - it needs at least `ndim+1` input points.  It doesn't fit in the current paradigm of `SphericalVoronoi` either, since the Voronoi regions of this input aren't described by a finite number of vertices.

pmla commented:
> What about equi-spaced generators on the equator? I think we may not handle the last case robustly yet anyway, but if it does produce spherical rectangles that wrap around the sphere it might be pathological.

I will implement this at some point.  I think the Voronoi regions will consist of two spherical triangles, each with two vertices on the equator and one at the north/south pole.  It don't think it will be pathological.

tylerjereddy commented:
I've tentatively added the 1.5.0 milestone as a follow-up to (hopefully) N-dimensional support in 1.4.x.

pmla commented:
This is a very nice test. I have added it in the most recent commit.

The problem was incorrect handling of single hemisphere generators on a geodesic. The previous code assumed that the generators lay in both hemispheres. I have replaced this ad-hoc calculation of Voronoi vertices on the geodesic by simply creating a lower dimensional SphericalVoronoi object. This code is also shorter.

tylerjereddy commented:
I tried compressing the regions down a bit with:

```diff
--- a/scipy/spatial/tests/test_spherical_voronoi.py
+++ b/scipy/spatial/tests/test_spherical_voronoi.py
@@ -384,14 +384,17 @@ class TestSphericalVoronoi(object):
         areas = sv.calculate_areas()
         assert_almost_equal(areas, 4 * np.pi / len(points))

-    def test_ultra_close_gens(self):
+
+    @pytest.mark.parametrize("arc_fraction", [
+        1.0, 1e-2, 1e-3])
+    def test_ultra_close_gens(self, arc_fraction):
         # use geometric_slerp to produce generators that
         # are close together, to push the limits
         # of the area (angle) calculations
         # also, limit generators to a single hemisphere
         path = geometric_slerp([0, 0, 1],
                                [1, 0, 0],
-                               t=np.linspace(0, 1, 1000))
+                               t=np.linspace(0, arc_fraction, 1000))
         sv = SphericalVoronoi(path)
         areas = sv.calculate_areas()
         assert_almost_equal(areas.sum(), 4 * np.pi)
```

That passes tests, which is good. If I try to squeeze the generators too close together it starts to register as a rank deficiency or another kind of degeneracy, which I suppose is to be expected.

tylerjereddy commented on scipy/spatial/tests/test_spherical_voronoi.py:
```diff
@@ -12,11 +12,70 @@
+    elif name == "icosahedron":
+        p = _generate_icosahedron()
+    else:
+        raise ValueError("unrecognized polyhedron")
```
This is the first line of code in SphericalVoronoi source or testing that isn't covered by a test. Since it is a test support function I didn't insist it should be covered, because then you're testing your tests, but I also wasn't sure the error handling was called for on a test support function & it can be nice to have 100 % coverage across the board.

Minor point though.

pmla replied:
I generate some polyhedra in spatial/transform/tests/test_rotation_groups.py` too, so there is some duplication of functionality here. We could collect the functions in a separate file if you think it is appropriate.

tylerjereddy commented:
@pmla Can you please add a release note to the `spatial` section: https://github.com/scipy/scipy/wiki/Release-note-entries-for-SciPy-1.5.0

pmla commented:
Thanks for the review and merge Tyler. I have updated the release notes.

pmla commented on scipy/spatial/_spherical_voronoi.py:
```diff
@@ -159,20 +183,18 @@ def __init__(self, points, radius=1, center=None, threshold=1e-06):
             raise ValueError("Radius inconsistent with generators.")
-        self.vertices = None
-        self.regions = None
-        self._tri = None
```
Thanks for the tip.

pmla commented on scipy/spatial/_spherical_voronoi.py:
```diff
@@ -236,24 +243,24 @@ def _calc_vertices_regions(self):
             return
 
         # get Convex Hull
-        self._tri = scipy.spatial.ConvexHull(self.points)
```
OK. My motivation is that the `self._tri` uses a lot of memory, and we only need the vertices and the simplices from it. The area calculation implementation needs a lot of memory too, so we can free a lot up by garbage collecting `self._tri`.

rgommers replied:
It's private, I don't see an issue with removing it.

rgommers replied:
Sorry, I missed the message the first time around.

pmla commented:
Slightly off-topic for this PR, but it has occurred to me that the Voronoi vertices for circular input are incorrect. Circular input has exactly two Voronoi vertices - one at each pole. The other "vertices" I have calculated (those in between each pair of adjacent input points) make it possible to plot regions correctly, but are not actually Voronoi vertices. If we use the actual Voronoi vertices only, we need a different mechanism for region plotting. This could be e.g. representing regions as plane normals, but this would be a big change. Otherwise we could add a note to explain that only two vertices are actual Voronoi vertices, and the rest are merely "helper" vertices for plotting.

tylerjereddy commented:
Maybe open a separate issue for that discussion? I haven't seen any literature on the circular input. Some might argue it is a case of degeneracy I suppose. `geometric_slerp()` is pretty useful for region plotting; can be used twice--once to find the arc midpoint of the adjacent generators and a second time to produce the Voronoi edges from poles through equatorial midpoints I suppose.
